src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -5096,6 +5096,426 @@ HCIMPL0(void, JIT_DebugLogLoopCloning)
 }
 HCIMPLEND
 
+#ifdef FEATURE_ON_STACK_REPLACEMENT
+
+// Runtime state for a patchpoint.
+//
+// Patchpoint identity is currently the return address
+// of the helper call in the jitted code.
+//
+// Todo: handle cases where IP addresses get recycled
```

Comment:
These should be fields in LoaderAllocator to get proper lifetime, that will also address the IL recycling problem.

Reply:
Have introduced an `OnStackReplacementManager` and reference it from the loader allocator; the patchpoint helper now maps from jitted code address to the right Manager instance.

Still not sure what memory pool the Manager and the table it refers to should be allocated from.

Reply:
`PerPatchpointInfo`s can be allocated on GetLowFrequencyHeap().

If they are allocated on a regular heap (like what it is in the current PR), they would need to be explicitly de-allocated in the destructor.

Reply:
Looks like eehash keeps its bucket table on the default heap. I guess I should use some other hash table type?

Reply:
I think it is fine to keep the eehash bucket table on the default heap.

Reply:
Also what do you recommend -- should I cache a pointer to the loader allocator (or the heap) in the  manager, or pass one or the other down as an arg to the GetPerPatchpoint method ...?

Reply:
I would keep a pointer to the owning LoaderAllocator in the manager.

Reply:
Thanks, updated accordingly.